README.md
# Php mpesa library

[![Latest Stable Version](https://poser.pugx.org/hackdelta/mpesa/v)](//packagist.org/packages/hackdelta/mpesa)
[![Total Downloads](https://poser.pugx.org/hackdelta/mpesa/downloads)](//packagist.org/packages/hackdelta/mpesa)
[![Latest Unstable Version](https://poser.pugx.org/hackdelta/mpesa/v/unstable)](//packagist.org/packages/hackdelta/mpesa)
[![License](https://poser.pugx.org/hackdelta/mpesa/license)](//packagist.org/packages/hackdelta/mpesa)
[![PHPUnit Status](https://github.com/Ghostscypher/php-mpesa/actions/workflows/php.yml/badge.svg?branch=main)](https://github.com/Ghostscypher/php-mpesa/actions?query=branch%3Amain)

## Table of contents

- [Php mpesa library](#php-mpesa-library)
  - [Table of contents](#table-of-contents)
  - [Introduction](#introduction)
  - [Getting Started](#getting-started)
  - [Installation](#installation)
  - [Usage](#usage)
    - [Initiating the library](#initiating-the-library)
    - [Editing configurations](#editing-configurations)
    - [Example usage for the whole library](#example-usage-for-the-whole-library)
    - [Register validation and confirmation urls (C2B)](#register-validation-and-confirmation-urls-c2b)
    - [Simulate a transaction (C2B)](#simulate-a-transaction-c2b)
    - [Initiate STK push (C2B)](#initiate-stk-push-c2b)
    - [Checking the stk push query status (C2B)](#checking-the-stk-push-query-status-c2b)
    - [Sending money (B2B)](#sending-money-b2b)
    - [Sending money (B2C)](#sending-money-b2c)
    - [Check transaction status (General)](#check-transaction-status-general)
    - [Request for reversal (General)](#request-for-reversal-general)
    - [Check for balance (General)](#check-for-balance-general)
    - [Register pull request callback URL (General)](#register-pull-request-callback-url-general)
    - [Send pull request (General)](#send-pull-request-general)
  - [Advanced usage](#advanced-usage)
  - [How To Contribute](#how-to-contribute)
  - [Getting Help](#getting-help)
  - [Security](#security)
  - [Testing](#testing)

## Introduction

A PHP Library that wraps around the mpesa APIs

## Getting Started

This library runs on the following:

- PHP ^7.4
- PHP ^8.0

## Installation

You can install the package via composer
`composer require hackdelta/mpesa`

## Usage

In order to check for recent changes kindly check the [Changelog](CHANGELOG.md)
The purpose of this package is to wrap around Mpesa APIs in order to provide an easier
and much neater methods. This library uses [Guzzle http library](https://github.com/guzzle/guzzle)
under the hood for http requests.

This documentation will be divided into five sections covering specific areas. Any other
additional notes are welcomed. Feel free to contribute.

1. Introduction to the library
2. C2B transactions
3. B2B transactions
4. B2C transactions
5. General transactions

For further details that might not have been mention here check:

1. [Library specifications](specifications.md)
2. [Mpesa endpoints](mpesa_endpoints.md)
3. [Safaricom developer portal](https://developer.safaricom.co.ke)

### Initiating the library

The library requires one to pass a configuration that will be used in various parts of the library

```php

use Hackdelta\Mpesa\Mpesa;
use Hackdelta\Mpesa\Main\MpesaConfig;
use Hackdelta\Mpesa\Extras\MpesaConstants;

$config = new MpesaConfig(
      [
         // API Credentials
         'consumer_key'    => 'YOUR_CONSUMER_KEY',
         'consumer_secret' => 'YOUR_CONSUMER_SECRET',

         // Environment
         'environment' => 'sandbox', // on sandbox environment
         // 'environment' => 'production', // on production

         // User credential
         'initiator_name'     => 'INITIATOR_NAME',
         'initiator_password' => 'INITIATOR_PASSWORD',

         // Or
         'security_credential' => '',

         // Used in combination with initiator name
         // and password
         // If you have different mpesa cert paths you can set them here 
         'sandbox_certificate_path'    => '',
         'production_certificate_path' => '',

         // Lipa na mpesa online passkey
         // For STK push
         'passkey' => 'PASSKEY',

         // Short code
         'short_code'          => 'SHORT_CODE', // This is your till or paybill
         'business_short_code' => 'BUSINESS_SHORT_CODE', // This is your head office number which for
                                                         // paybills it's the same, leave this blank
                                                         // if you are using paybill

         // Identifier type of the shortcode
         'identifier_type' => MpesaConstants::MPESA_IDENTIFIER_TYPE_PAYBILL, // For paybills
         // 'identifier_type' => MpesaConstants::MPESA_IDENTIFIER_TYPE_TILL, // For till

         // URLS
         // Webhook URLS
         'confirmation_url' => 'https://domain.example/confirm.php', // Where confirmation
                                                                     // callbacks will hit
         'validation_url'   => 'https://domain.example/validate.php',

         // STK callback URL
         'stk_callback_url' => 'https://domain.example/callback.php',

         // Query results URL
         'queue_timeout_url' => 'https://domain.example/callback.php',
         'result_url'        => 'https://domain.example/callback.php',

         // Pull request
         'organization_msisdn' => '0722000000', // Used for pull requests, this is the
                                                // number that receives confirmation messages
         'pull_callback_url'   => 'https://domain.example/callback.php',
      ]
   );

$mpesa = new Mpesa($config);

```

You don't have to set all those configurations, it's just listed here for all configuration
options available to you.

### Editing configurations

After initiation, each and every class provides you with a way of editing the configuration class
on the fly

For example

```php
use Hackdelta\Mpesa\Mpesa;
use Hackdelta\Mpesa\Main\MpesaConfig;

$config = new MpesaConfig(...); // Remember to pass in the configurations here
$mpesa = new Mpesa($config);

$mpesa->getConfig()
      ->setSandboxEnvironment(false) // Change environment to production, true means sandbox
      ...
      ->setConsumerKey('new consumer key');

```

**Please note that the set methods can be chained.**

For a full list of available methods check the [specifications](specifications.md#mpesa-configuration-class)

### Example usage for the whole library

This just shows an overview of a typical scenario this might be used for testing purposes, in this
example we use test credentials, you are however required to provide all the callback URLS, the
consumer key, the consumer secret, and your phone number for stk push test.
You can get the test credentials from [Safaricom developer portal](https://developers.safaricom.co.ke)
**Note: Some of this tests will fail because of having different shortcodes for B2B, anc C2B, in short B2B, C2B, reversal, and pull transaction requests are expected to fail**

```php
use Hackdelta\Mpesa\Mpesa;
use Hackdelta\Mpesa\Main\MpesaConfig;
use Hackdelta\Mpesa\Extras\MpesaConstants;
use Hackdelta\Mpesa\Exceptions\MpesaInternalException;
use Hackdelta\Mpesa\Exceptions\MpesaClientExceptions;
use Hackdelta\Mpesa\Exceptions\MpesaServerException;

$config = new MpesaConfig([
   // API Credentials
   'consumer_key'    => 'YOUR_CONSUMER_KEY', // <--- Change this
   'consumer_secret' => 'YOUR_CONSUMER_SECRET', // <--- Change this

   // Environment
   'environment' => 'sandbox', // on sandbox environment
   // 'environment' => 'production', // on production

   // User credential
   'initiator_name'     => 'testAPI497',
   'initiator_password' => 'Safaricom111!',

   // Lipa na mpesa online passkey
   // For STK push
   'passkey' => 'bfb279f9aa9bdbcf158e97dd71a467cd2e0c893059b10f78e6b72ada1ed2c919',

   // Short code
   'short_code'          => '601497', // This is your till or paybill
   'business_short_code' => '601497', // This is your head office number which for
                                                   // paybills it's the same, leave this blank
                                                   // if you are using paybill

   // Identifier type of the shortcode
   'identifier_type' => MpesaConstants::MPESA_IDENTIFIER_TYPE_PAYBILL, // For paybills
   // 'identifier_type' => MpesaConstants::MPESA_IDENTIFIER_TYPE_TILL, // For till

   // URLS
   // Webhook URLS
   'confirmation_url' => 'https://domain.example.com/confirm.php', // Where confirmation
                                                               // callbacks will hit
   'validation_url'   => 'https://domain.example.com/validate.php',

   // STK callback URL
   'stk_callback_url' => 'https://domain.example.com/callback.php',

   // Query results URL
   'queue_timeout_url' => 'https://domain.example.com/callback.php',
   'result_url'        => 'https://domain.example.com/callback.php',

   // Pull request
   'organization_msisdn' => '0722000000', // Used for pull requests, this is the
                                          // number that receives confirmation messages
   'pull_callback_url'   => 'https://domain.example.com/callback.php',

]);

$mpesa = new Mpesa($config);

// Check internal
function test_getting_auth_token() {
   global $mpesa;

   echo $mpesa->getConfig()->getAuth()->getToken();
}

// C2B
function test_registering_c2b_callbacks() {
   global $mpesa;

   echo $mpesa->C2B()->registerURL();
}

function test_simulating_transaction() {
   global $mpesa;

   echo $mpesa->C2B()->simulate('254708374149', 100, 'test account reference');
}

function test_initiating_stk_push() {
   global $mpesa;

   $mpesa->getConfig()
         ->setConfig("short_code", '174379')
         ->setConfig("business_short_code", '174379');
   
   // Change "254700000000" to your phone number
   echo $mpesa->C2B()->initiateSTKPush("254700000000", 1, "Test", "Test");

   // Reset config back
   $mpesa->getConfig()
         ->setConfig("short_code", '601497')
         ->setConfig("business_short_code", '601497');
   
}

function test_getting_Stk_push_status() {
   global $mpesa;
   
   $mpesa->getConfig()
         ->setConfig("short_code", '174379')
         ->setConfig("business_short_code", '174379');

   // Change the checkout id to the one you git from running the above
   echo $mpesa->C2B()->STKPushQuery("ws_CO_030520210342393444");

   // Reset config back
   $mpesa->getConfig()
         ->setConfig("short_code", '601497')
         ->setConfig("business_short_code", '601497');
}

// General
function test_getting_balance() {
   global $mpesa;

   echo $mpesa->checkBalance();
}

function test_reversal() {
   global $mpesa;

   // Change 'PE341HJ3Q8' to a test value gotten after simulating the transaction
   echo $mpesa->reverseTransaction('PE341HJ3Q8', 100, '254708263715', MpesaConstants::MPESA_IDENTIFIER_TYPE_MSISDN);

}

function test_checking_transaction_Status() {
   global $mpesa;

   echo $mpesa->checkTransactionStatus('PE341HJ3Q8');

}

function test_register_pull_url() {
   global $mpesa;

   echo $mpesa->pullRequestRegisterURL();
}

function test_pull() {
   global $mpesa;

   echo $mpesa->pullRequestQuery("2019-07-31 20:00:00", "2019-07-31 22:00:00");
}

// B2C
function test_B2C() {
   global $mpesa;

   echo $mpesa->B2C()->send(
        100, 
        '254700000000',
        MpesaConstants::MPESA_COMMAND_ID_SALARY_PAYMENT 
    );
}

// B2B
function test_B2B() {
   global $mpesa;

   echo $mpesa->B2B()->send(
        10, 
        '600000',
        MpesaConstants::MPESA_COMMAND_ID_BUSINESS_PAY_BILL,
        MpesaConstants::MPESA_IDENTIFIER_TYPE_PAYBILL,
        "account" 
    );
}

// In order to run the test uncomment one or more of the following
try {

   /**
    * This tests should pass provided you've set the correct configurations
   */

   /* Pre request check */
   test_getting_auth_token();

   /* C2B */
   // test_registering_c2b_callbacks();
   // test_simulating_transaction();
   // test_initiating_stk_push();
   // test_getting_Stk_push_status();

   /* B2B */
   // test_B2B(); // Expected to fail on sandbox

   /* B2C */
   // test_B2C(); // Expected to fail on sandbox

   /* General */
   // test_getting_balance();
   // test_checking_transaction_Status();
   // test_reversal(); // Expected to fail on sandbox
   // test_register_pull_url(); // Expected to fail on sandbox
   // test_pull(); // Expected to fail on sandbox

} catch(MpesaInternalException $e) {
   // This exception occurs when a configuration is missing e.g. consumer key
   echo $e->getMessage();

} catch(MpesaClientException $ce){
   // This occurs when the request has been sent to the mpesa servers but some invalid
   // data was supplied e.g. invalid consumer key
   
   echo $ce->getStatusCode(); // Returns http status code of the error
   echo $ce->getMessage(); // Returns the error message
   echo $ce->getErrorBody(); //  Returns the json string of the error body returned from server
   print_r($ce->getRequestParameters()); // An array of the request parameters sent i.e. the headers,
                                          // the request body, the URL hit, and method

} catch (MpesaServerException $se) {
   // This occurs when the request has been sent to the mpesa servers
   // but the server experienced some issues this is an error from the gateway and
   // is not the client fault

   echo $se->getStatusCode(); // Returns http status code of the error
   echo $se->getMessage(); // Returns the error message
   echo $se->getErrorBody(); //  Returns the json string of the error body returned from server
   print_r($se->getRequestParameters()); // An array of the request parameters sent i.e. the headers,
                                          // the request body, the URL hit, and method
}

```

### Register validation and confirmation urls (C2B)

To register the confirmation and validation urls you'll need to set the following configurations

**Note: This can be done only once in production***

```php
use Hackdelta\Mpesa\Mpesa;
use Hackdelta\Mpesa\Main\MpesaConfig;
use Hackdelta\Mpesa\Extras\MpesaConstants;
use Hackdelta\Mpesa\Exceptions\MpesaInternalException;
use Hackdelta\Mpesa\Exceptions\MpesaClientExceptions;
use Hackdelta\Mpesa\Exceptions\MpesaServerException;

$config = new MpesaConfig(
      [
         // API Credentials
         'consumer_key'    => 'YOUR_CONSUMER_KEY',
         'consumer_secret' => 'YOUR_CONSUMER_SECRET',

         // Environment
         'environment' => 'sandbox', // on sandbox environment
         
         // Short code
         'short_code'  => 'SHORT_CODE', // This is your till or paybill

         // Identifier type of the shortcode
         'identifier_type' => MpesaConstants::MPESA_IDENTIFIER_TYPE_PAYBILL, // For paybills
         // 'identifier_type' => MpesaConstants::MPESA_IDENTIFIER_TYPE_TILL, // For till


         // Webhook URLS
         'confirmation_url' => 'https://domain.example/confirm.php',
         'validation_url'   => 'https://domain.example/validate.php',
      
      ]
);

$mpesa = new Mpesa($config);

// Register urls

try{

   $response = $mpesa->C2B()->registerURL(); // Returns the MpesaResponse class object

   echo $response;
   // or
   echo $response->getJSONString();

} catch(MpesaInternalException $e) {
   // This exception occurs when a configuration is missing e.g. consumer key
   echo $e->getMessage();

} catch(MpesaClientException $ce){
   // This occurs when the request has been sent to the mpesa servers but some invalid
   // data was supplied e.g. invalid consumer key
   
   echo $ce->getStatusCode(); // Returns http status code of the error
   echo $ce->getMessage(); // Returns the error message
   echo $ce->getErrorBody(); //  Returns the json string of the error body returned from server
   print_r($ce->getRequestParameters()); // An array of the request parameters sent i.e. the headers,
                                          // the request body, the URL hit, and method

} catch (MpesaServerException $se) {
   // This occurs when the request has been sent to the mpesa servers
   // but the server experienced some issues this is an error from the gateway and
   // is not the client fault

   echo $se->getStatusCode(); // Returns http status code of the error
   echo $se->getMessage(); // Returns the error message
   echo $se->getErrorBody(); //  Returns the json string of the error body returned from server
   print_r($se->getRequestParameters()); // An array of the request parameters sent i.e. the headers,
                                          // the request body, the URL hit, and method

}

```

### Simulate a transaction (C2B)

This is used to simulate a transaction during sandbox testing, if used in production this will throw
an error

```php
use Hackdelta\Mpesa\Mpesa;
use Hackdelta\Mpesa\Main\MpesaConfig;
use Hackdelta\Mpesa\Extras\MpesaConstants;
use Hackdelta\Mpesa\Exceptions\MpesaInternalException;
use Hackdelta\Mpesa\Exceptions\MpesaClientExceptions;
use Hackdelta\Mpesa\Exceptions\MpesaServerException;

$config = new MpesaConfig(
      [
         // API Credentials
         'consumer_key'    => 'YOUR_CONSUMER_KEY',
         'consumer_secret' => 'YOUR_CONSUMER_SECRET',

         // Environment
         'environment' => 'sandbox', // on sandbox environment
         
         // Short code
         'short_code'  => 'SHORT_CODE', // This is your till or paybill

         // Identifier type of the shortcode
         'identifier_type' => MpesaConstants::MPESA_IDENTIFIER_TYPE_PAYBILL, // For paybills
         // 'identifier_type' => MpesaConstants::MPESA_IDENTIFIER_TYPE_TILL, // For til

      ]
);

$mpesa = new Mpesa($config);

// Simulate a transaction

try{

   $response = $mpesa->C2B()->simulate(
      '25470000000', // Phone number to simulate, use test credentials
      100, // Amount to transact
      'account number' // if simulating for paybill otherwise should be blank
   ); // Returns the MpesaResponse class object

   echo $response;
   // or
   echo $response->getJSONString();

} catch(MpesaInternalException $e) {
   // This exception occurs when a configuration is missing e.g. consumer key
   echo $e->getMessage();

} catch(MpesaClientException $ce){
   // This occurs when the request has been sent to the mpesa servers but some invalid
   // data was supplied e.g. invalid consumer key
   
   echo $ce->getStatusCode(); // Returns http status code of the error
   echo $ce->getMessage(); // Returns the error message
   echo $ce->getErrorBody(); //  Returns the json string of the error body returned from server
   print_r($ce->getRequestParameters()); // An array of the request parameters sent i.e. the headers,
                                          // the request body, the URL hit, and method

} catch (MpesaServerException $se) {
   // This occurs when the request has been sent to the mpesa servers
   // but the server experienced some issues this is an error from the gateway and
   // is not the client fault

   echo $se->getStatusCode(); // Returns http status code of the error
   echo $se->getMessage(); // Returns the error message
   echo $se->getErrorBody(); //  Returns the json string of the error body returned from server
   print_r($se->getRequestParameters()); // An array of the request parameters sent i.e. the headers,
                                          // the request body, the URL hit, and method

}
```

### Initiate STK push (C2B)

This is used for lipa na mpesa online transaction and requires the following configs

```php
use Hackdelta\Mpesa\Mpesa;
use Hackdelta\Mpesa\Main\MpesaConfig;
use Hackdelta\Mpesa\Extras\MpesaConstants;
use Hackdelta\Mpesa\Exceptions\MpesaInternalException;
use Hackdelta\Mpesa\Exceptions\MpesaClientExceptions;
use Hackdelta\Mpesa\Exceptions\MpesaServerException;

$config = new MpesaConfig(
      [
         // API Credentials
         'consumer_key'    => 'YOUR_CONSUMER_KEY',
         'consumer_secret' => 'YOUR_CONSUMER_SECRET',

         // Environment
         'environment' => 'sandbox', // on sandbox environment
         
         // Short code
         'short_code'  => 'SHORT_CODE', // This is your till or paybill
         'business_short_code' => 'BUSINESS_SHORT_CODE', // This is your head office number which for
                                                         // paybills it's the same, leave this blank
                                                         // if you are using paybill

         // Identifier type of the shortcode
         'identifier_type' => MpesaConstants::MPESA_IDENTIFIER_TYPE_PAYBILL, // For paybills
         // 'identifier_type' => MpesaConstants::MPESA_IDENTIFIER_TYPE_TILL, // For til

         // Lipa na mpesa online passkey
         // For STK push
         'passkey' => 'PASSKEY',

         // STK callback URL
         'stk_callback_url' => 'https://domain.example/callback.php',

      ]
);

$mpesa = new Mpesa($config);

// Initiate an STK push request

try{

   $response = $mpesa->C2B()->initiateSTKPush(
      '25470000000', // Phone number to send request to
      100, // Amount to transact

      // Optional parameters
      'account number', // The account reference
      'description', // Description accompanying the transaction 
      'timestamp' // Your own timestamp

   ); // Returns the MpesaResponse class object

   echo $response;
   // or
   echo $response->getJSONString();

} catch(MpesaInternalException $e) {
   // This exception occurs when a configuration is missing e.g. consumer key
   echo $e->getMessage();

} catch(MpesaClientException $ce){
   // This occurs when the request has been sent to the mpesa servers but some invalid
   // data was supplied e.g. invalid consumer key
   
   echo $ce->getStatusCode(); // Returns http status code of the error
   echo $ce->getMessage(); // Returns the error message
   echo $ce->getErrorBody(); //  Returns the json string of the error body returned from server
   print_r($ce->getRequestParameters()); // An array of the request parameters sent i.e. the headers,
                                          // the request body, the URL hit, and method

} catch (MpesaServerException $se) {
   // This occurs when the request has been sent to the mpesa servers
   // but the server experienced some issues this is an error from the gateway and
   // is not the client fault

   echo $se->getStatusCode(); // Returns http status code of the error
   echo $se->getMessage(); // Returns the error message
   echo $se->getErrorBody(); //  Returns the json string of the error body returned from server
   print_r($se->getRequestParameters()); // An array of the request parameters sent i.e. the headers,
                                          // the request body, the URL hit, and method

}
```

### Checking the stk push query status (C2B)

This is used to check the status of an stk push request

```php
use Hackdelta\Mpesa\Mpesa;
use Hackdelta\Mpesa\Main\MpesaConfig;
use Hackdelta\Mpesa\Extras\MpesaConstants;
use Hackdelta\Mpesa\Exceptions\MpesaInternalException;
use Hackdelta\Mpesa\Exceptions\MpesaClientExceptions;
use Hackdelta\Mpesa\Exceptions\MpesaServerException;

$config = new MpesaConfig(
      [
         // API Credentials
         'consumer_key'    => 'YOUR_CONSUMER_KEY',
         'consumer_secret' => 'YOUR_CONSUMER_SECRET',

         // Environment
         'environment' => 'sandbox', // on sandbox environment
         
         // Short code
         'short_code'  => 'SHORT_CODE', // This is your till or paybill
         'business_short_code' => 'BUSINESS_SHORT_CODE', // This is your head office number which for
                                                         // paybills it's the same, leave this blank
                                                         // if you are using paybill

         // Identifier type of the shortcode
         'identifier_type' => MpesaConstants::MPESA_IDENTIFIER_TYPE_PAYBILL, // For paybills
         // 'identifier_type' => MpesaConstants::MPESA_IDENTIFIER_TYPE_TILL, // For til

         // Lipa na mpesa online passkey
         // For STK push
         'passkey' => 'PASSKEY',

      ]
);

$mpesa = new Mpesa($config);


CHANGELOG.md
# Changelog

## Release Notes for 1.x

### Added

* Changelog


src/Mpesa.php
<?php

namespace Hackdelta\Mpesa;

use Hackdelta\Mpesa\Extras\MpesaConstants;
use Hackdelta\Mpesa\Extras\Validatable;
use Hackdelta\Mpesa\Main\MpesaB2B;
use Hackdelta\Mpesa\Main\MpesaB2C;
use Hackdelta\Mpesa\Main\MpesaC2B;
use Hackdelta\Mpesa\Main\MpesaConfig;
use Hackdelta\Mpesa\Main\MpesaHttp;
use Hackdelta\Mpesa\Main\MpesaResponse;

/**
 * The main class that will be instantiated.
 */
class Mpesa
{
    use Validatable;

    protected static ?MpesaHttp $http_client = null;

    protected static ?MpesaB2C $B2C = null;
    protected static ?MpesaC2B $C2B = null;
    protected static ?MpesaB2B $B2B = null;

    protected MpesaConfig $config;

    public function __construct(MpesaConfig $config)
    {
        $this->config = $config;

        if (self::$http_client === null) {
            self::$http_client = new MpesaHttp($this->config);
        }
    }

    public function setConfig(MpesaConfig $config): self
    {
        $this->config = $config;

        // Change the other classes config to
        $this->B2B()->setConfig($config);
        $this->B2C()->setConfig($config);
        $this->C2B()->setConfig($config);

        return $this;
    }

    public function getConfig(): MpesaConfig
    {
        return $this->config;
    }

    public function C2B(): MpesaC2B
    {
        return self::$C2B === null ?
            self::$C2B = new MpesaC2B($this->config) :
            self::$C2B;
    }

    public function B2B(): MpesaB2B
    {
        return self::$B2B === null ?
            self::$B2B = new MpesaB2B($this->config) :
            self::$B2B;
    }

    public function B2C(): MpesaB2C
    {
        return self::$B2C === null ?
            self::$B2C = new MpesaB2C($this->config) :
            self::$B2C;
    }

    // Account balance query
    // Optional remarks sent with the request
    public function checkBalance(string $remarks = 'remarks'): MpesaResponse
    {
        $url = sprintf(
            '%s%s',
            $this->config->getBaseURL(),
            MpesaConstants::MPESA_URIS['account_balance']
        );

        // Validate that data is correct
        $this->validateString('initiator_name', $this->config->getInitiator());
        $this->validateString('security_credential', $this->config->getSecurityCredential());
        $this->validateString('short_code', $this->config->getShortCode());
        $this->validateString('queue_timeout_url', $this->config->getQueueTimeoutURL());
        $this->validateString('result_url', $this->config->getResultURL());

        $this->validateString('remarks', $remarks);

        $response = self::$http_client->request(
            $url,
            'POST',
            [
                'Initiator'          => $this->config->getInitiator(),
                'SecurityCredential' => $this->config->getSecurityCredential(),
                'CommandID'          => MpesaConstants::MPESA_COMMAND_ID_ACCOUNT_BALANCE,
                'PartyA'             => $this->config->getShortCode(),
                'IdentifierType'     => $this->config->getIdentifierType(),
                'Remarks'            => $remarks,
                'QueueTimeOutURL'    => $this->config->getQueueTimeoutURL(),
                'ResultURL'          => $this->config->getResultURL(),
            ],
            [
                'Authorization' => sprintf('Bearer %s', $this->config->getAuth()->getToken()),
            ]
        );

        return $response;
    }

    // Transaction status, check the transaction status of an mpesa code
    // transaction_id: This is the mpesa code
    // Remarks: comments sent along with the transaction
    // Occasion: additional data sent with the transaction
    public function checkTransactionStatus(
        string $transaction_id,
        string $remarks = 'remarks',
        string $occasion = ' '
    ): MpesaResponse {
        $url = sprintf(
            '%s%s',
            $this->config->getBaseURL(),
            MpesaConstants::MPESA_URIS['transaction_status']
        );

        // Validate that data is correct
        $this->validateString('initiator_name', $this->config->getInitiator());
        $this->validateString('security_credential', $this->config->getSecurityCredential());
        $this->validateString('transaction_id', $transaction_id);
        $this->validateString('short_code', $this->config->getShortCode());
        $this->validateString('queue_timeout_url', $this->config->getQueueTimeoutURL());
        $this->validateString('result_url', $this->config->getResultURL());

        $this->validateString('remarks', $remarks);

        $response = self::$http_client->request(
            $url,
            'POST',
            [
                'Initiator'          => $this->config->getInitiator(),
                'SecurityCredential' => $this->config->getSecurityCredential(),
                'CommandID'          => MpesaConstants::MPESA_COMMAND_ID_TRANSACTION_STATUS_QUERY,
                'TransactionID'      => $transaction_id,
                'PartyA'             => $this->config->getShortCode(),
                'IdentifierType'     => $this->config->getIdentifierType(),
                'Remarks'            => $remarks,
                'QueueTimeOutURL'    => $this->config->getQueueTimeoutURL(),
                'ResultURL'          => $this->config->getResultURL(),
                'Occasion'           => $occasion,
            ],
            [
                'Authorization' => sprintf('Bearer %s', $this->config->getAuth()->getToken()),
            ]
        );

        return $response;
    }

    // Initiate reversal request
    // transaction_id: The Mpesa code.
    // amount: The amount that is being reversed
    // receiver_party: the shortcode, or MSISDN that received the payment
    // receiver_identifier_type: constant that shows the receiver identifier type
    //         possible values are; MPESA_IDENTIFIER_TYPE_MSISDN, MPESA_IDENTIFIER_TYPE_TILL,
    //                MPESA_IDENTIFIER_TYPE_PAYBILL , MPESA_IDENTIFIER_TYPE_SHORTCODE
    public function reverseTransaction(
        string $transaction_id,
        int $amount,
        string $receiver_party,
        string $receiver_identifier_type,
        string $remarks = 'remarks',
        string $occasion = ''
    ): MpesaResponse {
        $url = sprintf(
            '%s%s',
            $this->config->getBaseURL(),
            MpesaConstants::MPESA_URIS['reversal']
        );

        // Validate that data is correct
        $this->validateString('initiator_name', $this->config->getInitiator());
        $this->validateString('security_credential', $this->config->getSecurityCredential());
        $this->validateString('transaction_id', $transaction_id);
        $this->validateInt('amount', $amount, 1);
        $this->validateString('reciever_party', $receiver_party);

        $this->validateArray('reciever_identifier_type', $receiver_identifier_type, [
            MpesaConstants::MPESA_IDENTIFIER_TYPE_MSISDN,
            MpesaConstants::MPESA_IDENTIFIER_TYPE_PAYBILL,
            MpesaConstants::MPESA_IDENTIFIER_TYPE_TILL,
            MpesaConstants::MPESA_IDENTIFIER_TYPE_SHORTCODE,
        ]);

        $this->validateString('short_code', $this->config->getShortCode());
        $this->validateString('queue_timeout_url', $this->config->getQueueTimeoutURL());
        $this->validateString('result_url', $this->config->getResultURL());

        $this->validateString('remarks', $remarks);

        $response = self::$http_client->request(
            $url,
            'POST',
            [
                'Initiator'              => $this->config->getInitiator(),
                'SecurityCredential'     => $this->config->getSecurityCredential(),
                'CommandID'              => MpesaConstants::MPESA_COMMAND_ID_TRANSACTION_REVERSAL,
                'TransactionID'          => $transaction_id,
                'PartyA'                 => $this->config->getShortCode(),
                'Amount'                 => $amount,
                'IdentifierType'         => $this->config->getIdentifierType(),
                'ReceiverParty'          => $receiver_party,
                'RecieverIdentifierType' => $receiver_identifier_type,
                'Remarks'                => $remarks,
                'QueueTimeOutURL'        => $this->config->getQueueTimeoutURL(),
                'ResultURL'              => $this->config->getResultURL(),
                'Occasion'               => $occasion,
            ],
            [
                'Authorization' => sprintf('Bearer %s', $this->config->getAuth()->getToken()),
            ]
        );

        return $response;
    }

    /**
     * Register pull request endpoint.
     */
    public function pullRequestRegisterURL(): MpesaResponse
    {
        $url = sprintf(
            '%s%s',
            $this->config->getBaseURL(),
            MpesaConstants::MPESA_URIS['pull_transaction_register']
        );

        // Validate that data is correct
        $this->validateString('short_code', $this->config->getShortCode());
        $this->validateString('nominated_number', $this->config->getOrganizationMSISDN());
        $this->validateString('pull_callback_url', $this->config->getPullCallbackURL());

        $response = self::$http_client->request(
            'POST',
            $url,
            [
                'ShortCode'       => $this->config->getShortCode(),
                'RequestType'     => MpesaConstants::MPESA_REQUEST_TYPE_PULL,
                'NominatedNumber' => $this->config->getOrganizationMSISDN(),
                'CallBackURL'     => $this->config->getPullCallbackURL(),
            ],
            [
                'Authorization' => sprintf('Bearer %s', $this->config->getAuth()->getToken()),
            ]
        );

        return $response;
    }

    /**
     * Perform a pull request query.
     *
     * @param string $start_date: The start period of the missing transactions in the
     *                            format of 2019-07-31 20:35:21 or 2019-07-31 19:00
     * @param string $end_date:   The end of the period for the missing transactions in the
     *                            format of 2019-07-31 20:35:21 or 2019-07-31 22:35
     * @param string $offset:     Starts from 0. The service uses offset as opposed to page numbers.
     *                            The OFF SET value allows you to specify which row to start from retrieving
     *                            data. Suppose you wanted to show results 101-200. With the
     *                            OFFSET keyword you type the (page number/index/offset value) 100.
     */
    public function pullRequestQuery(string $start_date, string $end_date, int $offset = 0): MpesaResponse
    {
        $url = sprintf(
            '%s%s',
            $this->config->getBaseURL(),
            MpesaConstants::MPESA_URIS['pull_transaction_query']
        );

        // Validate that data is correct
        $this->validateString('short_code', $this->config->getShortCode());
        $this->validateString('start_date', $start_date);
        $this->validateString('end_date', $end_date);
        $this->validateInt('offset', $offset, 0);

        $response = self::$http_client->request(
            $url,
            'POST',
            [
                'ShortCode'   => $this->config->getShortCode(),
                'StartDate'   => $start_date,
                'EndDate'     => $end_date,
                'OffSetValue' => $offset,
            ],
            [
                'Authorization' => sprintf('Bearer %s', $this->config->getAuth()->getToken()),
            ]
        );

        return $response;
    }
}
